File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isGood(self, root: TreeNode, maxPrec:int) -> bool:
        return root.val >= maxPrec

    def goodNotesRec(self, root: TreeNode, maxPrec) -> int:
        if root == None:
            return 0
        currentGood = self.isGood(root, maxPrec)
        newMax = max(maxPrec, root.val)
        compteGauche = self.goodNotesRec(root.left, newMax)
        compteDroit = self.goodNotesRec(root.right, newMax)
        return currentGood + compteGauche + compteDroit
    
    def goodNodes(self, root: TreeNode) -> int:
        return self.goodNotesRec(root, float('-inf'))

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void levelOrderLevel(TreeNode* root, int level, vector<vector<int>>& mainVector) {
        if (root == nullptr) {
            return;
        }
        int taille = mainVector.size();
        if (taille > level) {
            mainVector[level].push_back(root->val);
        } else {
            mainVector.push_back(vector<int> {root->val});
        }
        levelOrderLevel(root->left, level + 1, mainVector);
        levelOrderLevel(root->right, level + 1, mainVector);
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> mainVector;
        levelOrderLevel(root, 0, mainVector);
        return mainVector;
    }
};